Generate DTS `wrangler-overrides.d.ts` from config

JacobMGEvans opened the pull request:
The type command aggregates the bindings from config and the rules for modules, then generates a DTS file for both service workers `declare global { ... }` or the preferred module workers `interface Env { ... }`

The rules for modules generates `declare module` based on the module type, the 3 types `Text`, `Data` & `CompiledWasm`
Module Example Outputs:

**CompiledWasm**
```ts
declare module "**/*.wasm" {
const value: WebAssembly.Module;
export default value;
}
``` 

**Data** 
```ts
declare module "**/*.webp" {
const value: ArrayBuffer;
export default value;
}
``` 
**Text**
```ts
declare module "**/*.text" {
const value: string;
export default value;
}
``` 

resolves #2034 
resolves #2033

JacobMGEvans commented on packages/wrangler/src/__tests__/type-generation.test.ts:
```diff
@@ -0,0 +1,218 @@
+				compatibility_date: "2022-01-12",
+				name: "test-name",
+				main: "./index.ts",
+				// @ts-expect-error This type is out of sync with the actual bindingsConfig type
+				vars: bindingsConfigMock.vars,
```
Its testing mock data... I can do that but the issue is deeper which is we have conflicting types between Config & CFInit

penalosa replied:
Can we fix that? That seems like a deeper issue that might be worth addressing?

JacobMGEvans commented on packages/wrangler/src/__tests__/type-generation.test.ts:
```diff
@@ -0,0 +1,218 @@
+				compatibility_date: "2022-01-12",
+				name: "test-name",
+				main: "./index.ts",
+				// @ts-expect-error This type is out of sync with the actual bindingsConfig type
+				vars: bindingsConfigMock.vars,
```
As above too lol

JacobMGEvans commented on packages/wrangler/src/__tests__/type-generation.test.ts:
```diff
@@ -0,0 +1,218 @@
+				compatibility_date: "2022-01-12",
+				name: "test-name",
+				main: "./index.ts",
+				// @ts-expect-error This type is out of sync with the actual bindingsConfig type
+				unsafe: bindingsConfigMock.unsafe?.at(0) ?? {},
```
Same.

JacobMGEvans commented on packages/wrangler/src/index.tsx:
```diff
@@ -467,6 +470,41 @@ export function createCLIParser(argv: string[]) {
 	);
 
+	// type generation
+	wrangler.command(
+		"types",
+		"📝  Generate types from bindings & module rules in config",
+		() => {},
+		async () => {
+			await printWranglerBanner();
+			const config = readConfig(undefined, {});
+
+			// Currently includes bindings & rules for declaring modules
+			type PartialConfigToDTS = CfWorkerInit["bindings"] & {
+				rules: Config["rules"];
+			};
+			const configBindings: PartialConfigToDTS = {
+				kv_namespaces: config.kv_namespaces || [],
+				vars: { ...config.vars },
+				wasm_modules: config.wasm_modules,
+				text_blobs: {
+					...config.text_blobs,
+				},
+				data_blobs: config.data_blobs,
+				durable_objects: config.durable_objects,
+				r2_buckets: config.r2_buckets,
+				d1_databases: identifyD1BindingsAsBeta(config.d1_databases),
+				services: config.services,
+				dispatch_namespaces: config.dispatch_namespaces,
+				logfwdr: config.logfwdr,
+				unsafe: config.unsafe?.bindings,
+				rules: config.rules,
+			};
+
+			await generateTypes(configBindings, config);
+		}
+	);
```
I think that is a bit of a hasty abstraction, its pretty simple, mostly an object literal being constructed and few functions.

JacobMGEvans commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,170 @@
+import type { CfWorkerInit } from "./worker";
+
+// Currently includes bindings & rules for declaring modules
+type PartialConfigToDTS = CfWorkerInit["bindings"] & {
+	rules: Config["rules"];
+};
```
Yeah... I just wasn't sure which place in belonged lol

JacobMGEvans commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,170 @@
+	if (configToDTS.vars) {
+		for (const varName in configToDTS.vars) {
+			const varValue = configToDTS.vars[varName];
+			if (typeof varValue === "string") {
```
Which are coerced to strings IIRC

JacobMGEvans commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,170 @@
+	let modulesTypeStructure = "";
+	if (configToDTS.rules) {
+		for (const ruleObject of configToDTS.rules) {
+			if (ruleObject.type === "Text") {
```
I think a refactor can occur later, I figured some sort of `switch` would make sense. Its non-blocking and stylistic coding change though.

JacobMGEvans replied:
Implemented suggestion.

penalosa commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,170 @@
+
+	if (configToDTS.kv_namespaces) {
+		for (const kvNamespace of configToDTS.kv_namespaces) {
+			envTypeStructure += `${kvNamespace.binding}: KVNamespace;
+      `;
```
I think a join makes most sense, but another alternative is to use literal `\n` characters in the string (I know template strings can be multi line, but explicit new lines would probably be clearer in this case)

JacobMGEvans replied:
> I think a join makes most sense, but another alternative is to use literal `\n` characters in the string (I know template strings can be multi line, but explicit new lines would probably be clearer in this case)

I ended up doing a combination of both

penalosa commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,171 @@
+	const entry = await getEntry({}, config, "types");
+	const envTypeStructure: string[] = [];
+
+	if (configToDTS.kv_namespaces) {
```
If you give the config default values for `kv_namespaces` etc... of an empty array, you could remove all these ifs

JacobMGEvans replied:
I will consider it.

penalosa commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,171 @@
+
+	const modulesTypeStructure: string[] = [];
+	if (configToDTS.rules) {
+		const typeToTypeScriptType = {
```
Something like `moduleTypeMap`?

JacobMGEvans replied:
Both mean the same thing?

penalosa replied:
Yes, but the word "type" in a variable 3 times is a bit confusing

JacobMGEvans replied:
Fair point

penalosa commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,171 @@
+		for (const ruleObject of configToDTS.rules) {
+			const typeScriptType =
+				typeToTypeScriptType[
+					ruleObject.type as keyof typeof typeToTypeScriptType
```
What if it's not `Data`, `Text`, or `CompiledWasm`?

JacobMGEvans replied:
Those are not handled. I am not sure what you are getting at?

penalosa replied:
That's my point though—where do we filter out modules of other types?

JacobMGEvans replied:
Why would we "filter them out" the table only has the 3 Types if the Type isn't on the table then this `typeScriptType moduleTypeMap[ruleObject.type as keyof typeof moduleTypeMap];`  wont map to a Type

JacobMGEvans replied:
Unless I am misunderstanding something 🤷🏻‍♂️

penalosa replied:
Ah, I missed line 105—apologies

penalosa commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,171 @@
+		typesString: string[],
+		formatType: "modules" | "service-worker"
+	) {
+		const wranglerOverrideDTSPath = findUpSync("wrangler-overrides.d.ts");
```
Bikeshedding, but can we call this something different? Maybe `configuration.d.ts`? Or `cloudflare.d.ts`?

JacobMGEvans replied:
Definitely not `configuration.d.ts` that is far too generic. Why `cloudflare.d.ts` over `wrangler-overrides.d.ts`?

penalosa replied:
Because they aren't overrides (what would they be overriding?). Maybe `worker-configuration.d.ts`?

JacobMGEvans replied:
Hmmm, yeah it's not really an override. I like this suggestion `worker-configuration.d.ts`

penalosa commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,171 @@
+
+		let combinedTypeStrings = "";
+		if (formatType === "modules") {
+			combinedTypeStrings =
```
Can this be a template string with explicit `\n` rather than `""+""`?

JacobMGEvans replied:
This seems like a stylistic suggestion, is there a benefit to making the change?

penalosa replied:
It's entirely stylistic, but I think it would significantly improve readability

JacobMGEvans replied:
Done.

penalosa commented on packages/wrangler/src/type-generation.ts:
```diff
@@ -0,0 +1,171 @@
+		if (envTypeStructure.length || modulesTypeStructure.length) {
+			fs.writeFileSync(
+				"wrangler-overrides.d.ts",
+				"// ***AUTO GENERATED BY WORKERS CLI WRANGLER***" +
```
Maybe something like `Generated by Wrangler on ${date}`?

JacobMGEvans replied:
Oh, I like this.

penalosa reviewed (changes requested):
Looking good! This'll be amazing to have

penalosa commented:
Also @JacobMGEvans the force pushing means it's impossible to use the "changes since last review" view on GitHub (unless there's some workaround for that)

JacobMGEvans commented on .changeset/long-buttons-retire.md:
````diff
@@ -0,0 +1,38 @@
+**CompiledWasm**
+
+```ts
+declare module "**/*.wasm" {
````
Sounds good, figured it just used the glob pattern.

JacobMGEvans commented on packages/wrangler/src/__tests__/type-generation.test.ts:
```diff
@@ -0,0 +1,217 @@
+				name: "test-name",
+				main: "./index.ts",
+				// eslint-disable-next-line @typescript-eslint/no-explicit-any
+				...(bindingsConfigMock as any),
```
Its mock data and the fundamental issue is with the type differences between types `Config` & `CfWorkerInit`

penalosa replied:
Can we document that difference somewhere? I'm not quite sure I understand the issue

JacobMGEvans replied:
I'm not really certain it's a good use of time to document type hacks in mock data.

penalosa replied:
I think I must be misunderstanding: Are `CfWorkerInit` and `Config` only used for mock data?

JacobMGEvans commented on packages/wrangler/src/__tests__/type-generation.test.ts:
```diff
@@ -0,0 +1,217 @@
+				compatibility_date: "2022-01-12",
+				name: "test-name",
+				main: "./index.ts",
+				// @ts-expect-error This type is out of sync with the actual bindingsConfig type
```
Its mock data and the fundamental issue is with the type differences between types Config & CfWorkerInit

JacobMGEvans commented on packages/wrangler/src/index.tsx:
```diff
@@ -467,6 +469,38 @@ export function createCLIParser(argv: string[]) {
+			const config = readConfig(undefined, {});
+
+			const configBindings: PartialConfigToDTS = {
+				kv_namespaces: config.kv_namespaces || [],
```
> Also @JacobMGEvans the force pushing means it's impossible to use the "changes since last review" view on GitHub (unless there's some workaround for that)

You technically can look at the commit's between the forces, it usually diffs them.

JacobMGEvans replied:
We should look at changing it from the code I pulled this from too then:
https://github.com/cloudflare/wrangler2/blob/965ab19b6a2fca6ef16928baadc3436677d78c80/packages/wrangler/src/publish/publish.ts#L482-L505

JacobMGEvans commented:
> Also @JacobMGEvans the force pushing means it's impossible to use the "changes since last review" view on GitHub (unless there's some workaround for that)

You technically can look at the commit's between the forces, it usually diffs them.
